### PR TITLE
Updated Support For App Indexing

### PR DESCRIPTION
**Pull request by sojanpr**
Supporting new Firebase App indexing for listing BUOs in google search. Only publicly indexable BUO’s are indexed on google.

The Firebase app app indexing dependency will be made optional when deploying the artifact

@aaustin @Sarkar

### CODE REVIEW

File: Branch-SDK/build.gradle [hunk 1]
```diff
@@ -7,12 +7,16 @@ dependencies {
     compile fileTree(dir: 'libs', include: '*.jar')
     compile 'com.android.support:support-annotations:22.2.0'
     compile 'com.crashlytics.sdk.android:answers-shim:0.0.6'
+
+    // This is an optional dependency for the SDK and is excluded while deploying. This dependency is to support
+    // Firebase app indexing if the target has Firebase App indexing enabled. Please check the gradle-mvn-push script to make sure this is added as an optional dependency
+    compile 'com.google.firebase:firebase-appindexing:10.0.1'
```
Comment (aaustin) [inline Branch-SDK/build.gradle]: Let's make this a bit clearer. I'd suggest the following comment:
```Please note that the Branch SDK does not require Firebase to operate. This dependency is listed here so there will not be build errors, but the library is *not* added to your app unless you do so yourself. Please check the code in gradle-mvn-push script to see how this works```

File: Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java [hunk 2]
```diff
@@ -17,80 +24,116 @@
...
-                        listOnGoogleSearch(urlForAppIndexing, context, buo);
+                        if (fireBaseUserActionsInstance != null) {
+                            addToAppIndexUsingFireBase(contentUrl, buo);
+                        } else {
+                            //noinspection deprecation
+                            listOnGoogleSearch(contentUrl, context, buo);
+                        }
                     } catch (Throwable e) {
-                        PrefHelper.Debug("BranchSDK", "Branch Warning: Unable to list your content in google search. Please make sure you have the latest google play libs added to your project");
+                        PrefHelper.Debug("BranchSDK", "Branch Warning: Unable to list your content in google search. Please make sure you have added latest firebase app indexing SDK to your project dependencies.");
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java]: Please capitalize pronouns (Google, Firebase) int his comment.

File: Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java [hunk 3]
```diff
@@ -17,80 +24,116 @@
...
                     }
                 }
             }
         }).run();
     }
-
+    
+    private static void addToAppIndexUsingFireBase(String contentUrl, BranchUniversalObject buo) {
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java]: base is lower case B in Firebase

File: Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java [hunk 4]
```diff
@@ -17,80 +24,116 @@
...
+        Indexable noteToIndex = Indexables.newSimple(buo.getTitle(), contentUrl);
+        Task<Void> task = FirebaseAppIndex.getInstance().update(noteToIndex);
+        
+        Action action = new Action.Builder(Action.Builder.VIEW_ACTION)
+                .setObject(buo.getTitle(), contentUrl)
+                .setMetadata(new Action.Metadata.Builder()
+                        .setUpload(buo.isPublicallyIndexable()))
+                .build();
+        
+        fireBaseUserActionsInstance.end(action);
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java]: same here with regard to capital B

File: Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java [hunk 5]
```diff
@@ -17,80 +24,116 @@
  * </p>
  */
 class AppIndexingHelper {
-    public static void addToAppIndex(final Context context, final BranchUniversalObject buo) {
+    private static FirebaseUserActions fireBaseUserActionsInstance = null;
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/indexing/AppIndexingHelper.java]: B here

File: README.md [hunk 6]
```diff
@@ -626,14 +626,14 @@ branchUniversalObject.generateShortUrl(this, linkProperties, new BranchLinkCreat
...
 
 ```java
 branchUniversalObject.listOnGoogleSearch(context);
 ```
-Note : Please make sure you have added the following to your gradle dependency
+Note : Please make sure Firebase app indexing is enabled for your application and has Firebase app indexing dependency added
 ```java
-'com.google.android.gms:play-services-appindexing:9.+'
+'compile 'com.google.firebase:firebase-appindexing:10.0.1'
 ```
```
Comment (aaustin) [inline README.md]: Maybe you can add a note about how to test whether the content was indexed?

Comment (aaustin) [inline README.md]: Also, we should make sure we update our official docs https://github.com/BranchMetrics/documentation/blob/master/pages/marketing-channels/google-app-indexing.md

**Comment by sojanpr** [comment]
Testing Steps
1. Enable Firebase on TestBed application
     Using Android Studio 2.3+ got to `Tools` -> `Firebase` open FirebaseAssistant. Enable App indexing for `TestBed App` from FirebaseAssistant

2. Update BUO in `MaiActivity.java` with test title , description and other properties
3. Press `Report View` Button.
4. Local Indexing : The content should be viewable in `In App ` search when searched  by title or description
5. Google Search Indexing : Get the indexed link from logs (Run app in test mode) and use this  link to verify if indexed https://firebase.google.com/docs/app-indexing/android/test#test-public-content-indexing
6.Control Indexing Test : 
   a) BUO will be locally indexed only if local index mode is `public`. Please use `setLocalIndexMode` builder method  to test this (Add  builder method `setLocalIndexMode ` to the  BUO in `MainActivity.java`)
b) BUO will be publicly indexed on Google search if index mode is public. Please use `setIndexMode` builder method  to test this (Add  builder method `setIndexMode ` to the  BUO in `MainActivity.java`)

**Comment by aaustin** [comment]
Looks good 👍

**Comment by Sarkar** [comment]
Can we get this merged in?

**Comment by sojanpr** [comment]
QA in progress. Will merge once it pass QA cycle

**Pull request merged**